academic-observatory-workflows/academic_observatory_workflows/openalex_telescope/tasks.py

```diff
@@ -0,0 +1,683 @@
+                logging.info(f"Error generating schema for file {input_path}: {schema_error}")
+
+            # Merge the schemas from each process. Each data file could have more fields than others.
+            merged_schema_map = merge_schema_maps(to_add=schema_map, old=merged_schema_map)
+
+    # Flatten schema from nested OrderedDicts to a regular Bigquery schema.
+    merged_schema = flatten_schema(schema_map=merged_schema_map)
+
+    # Save schema to file
+    with open(entity.generated_schema_path, mode="w") as f_out:
+        json.dump(merged_schema, f_out, indent=2)
+
+
+def upload_schema(*, entity: OpenAlexEntity, transform_bucket: str):
+    success = gcs_upload_files(bucket_name=transform_bucket, file_paths=[entity.generated_schema_path])
+    if not success:
+        raise AirflowException("upload_schema: error uploading schema")
+
+
+def compare_schemas(*, entity: OpenAlexEntity, transform_bucket: str, slack_conn_id: str, **context):
+    logging.info(f"Loading schemas from file: generated: {entity.generated_schema_path}")
+    logging.info(f"Expected: {entity.schema_file_path}")
+
+    # Download merged schema
+    success = gcs_download_blob(
+        bucket_name=transform_bucket,
+        blob_name=gcs_blob_name_from_path(entity.generated_schema_path),
+        file_path=entity.generated_schema_path,
+    )
+    if not success:
+        raise AirflowException("compare_schemas: error downloading schema")
+
+    # Read in the expected schema for the entity.
+    merged_schema = load_json(entity.generated_schema_path)
+    expected_schema = load_json(entity.schema_file_path)
+
+    try:
+        match = bq_compare_schemas(expected_schema, merged_schema, check_types_match=False)
+    except:
+        match = False
+
+    if not match:
+        logging.info("Generated schema and expected do not match! - Sending a notification via Slack")
+        slack_msg = f"Found differences in the OpenAlex entity {entity.entity_name} data structure for the data dump vs pre-defined Bigquery schema. Please investigate."
+        ti: TaskInstance = context["ti"]
+        execution_date = context["execution_date"]
+        send_slack_msg(ti=ti, logical_date=execution_date, comments=slack_msg, slack_conn_id=slack_conn_id)
+
+
+def upload_files(*, entity: OpenAlexEntity, transform_bucket: str):
+    # Make files to upload
+    file_paths = []
+    for entry in entity.entries:
+        file_path = os.path.join(entity.transform_folder, entry.object_key)
+        file_paths.append(file_path)
+
+    # Upload files
+    success = gcs_upload_files(bucket_name=transform_bucket, file_paths=file_paths)  # cloud_workspace.transform_bucket
+    if not success:
+        raise AirflowException("upload_files: error uploading files to cloud storage")
+
+
+def bq_load_table(*, entity: OpenAlexEntity):
+    table_name = "main"
+    table_id = entity.bq_table_id
+    description = entity.table_description
+
+    if bq.bq_table_exists(table_id):
```

Comment:
Could you set the write disposition to WRITE_EMPTY? Then you wouldn't need to put this check in.

Reply:
That's a good point, I've changed that.